Trail Map: implement the plans page subheader

southp opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #89788, based on #89800 

## Proposed Changes

This PR implements the Trail Map version of the plans page subheader:

Desktop:
<img width="1146" alt="CleanShot 2024-04-26 at 10 51 50@2x" src="https://github.com/Automattic/wp-calypso/assets/1842898/ae56f7a4-cac7-4ba8-8b43-53d62b9886b2">

Mobile:
<img width="590" alt="CleanShot 2024-04-26 at 10 52 43@2x" src="https://github.com/Automattic/wp-calypso/assets/1842898/f918536f-28f9-4f25-b8e1-be49790fded5">

When the Free plan is deemphasized, the subheader will be replaced by the same Free plan CTA header:

<img width="1063" alt="CleanShot 2024-04-26 at 10 53 53@2x" src="https://github.com/Automattic/wp-calypso/assets/1842898/d5d21513-6207-4e7f-8d76-f88ddf390414">

The design can be found in Uzd0AYiDAV4eFq1V35T9Ef-fi-673_7146. Note that we are not aiming at pixel-perfect implementation here since it will be an experiment. We can first get the structure correct, and then iterate on the details when the time budget allows.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Enable the feature flag by appending `?flags=onboarding/trail-map-feature-grid`
* Go to `/start/plans` and confirm that the new subheader shows correctly, on both destkop & mobile.
* The new subheader shouldn't show in the logged-in `/plans` and any other signup flows.
* Hard-code `deemphasizeFreePlan` as true here: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/index.tsx#L405, and confirm that the Free plan CTA subheader shows.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aneeshd16 commented on client/my-sites/plans-features-main/components/plans-page-subheader.tsx:
```diff
@@ -0,0 +1,134 @@
+
+	return (
+		<HeaderContainer>
+			<PrefixSection>{ translate( 'All plans includes:' ) }</PrefixSection>
```
```suggestion
			<PrefixSection>{ translate( 'All plans include:' ) }</PrefixSection>
```

Small fix. :)

aneeshd16 commented on client/my-sites/plans-features-main/components/plans-page-subheader.tsx:
```diff
@@ -0,0 +1,134 @@
+			<PrefixSection>{ translate( 'All plans includes:' ) }</PrefixSection>
+			<FeatureSection>
+				{ translate(
+					'{{Checkmark}}{{/Checkmark}}Website Building{{Checkmark}}{{/Checkmark}}Hosting{{Checkmark}}{{/Checkmark}}eCommerce',
+					{
+						components: {
+							Checkmark: <Gridicon icon="checkmark" size={ 18 } />,
+						},
+						comment: 'Checkmark is an icon showing a green check mark.',
+					}
+				) }
```
I am wondering if we can separate out the strings and the icons to make this easier to translate (perhaps these strings are already translated?)
```jsx
<div style="display: flex">
    <span><Gridicon icon="checkmark" size={ 18 } />{ translate( 'Website Building' ) }</span>
    <span><Gridicon icon="checkmark" size={ 18 } />{ translate( 'Hosting' ) }</span>
    <span><Gridicon icon="checkmark" size={ 18 } />{ translate( 'eCommerce' ) }</span>
</div>

```

southp committed:
Extract the subheader logic into the PlansPageSubheader component

Scaffolding the structure of showing the Trail Map version of the
subheader

Restructure the subheader for the mobile design need.

Fine tune the layout style

Adjusting the margins accordingly

Fix the mobile layout

Reuse styles from Subheader and move the styled components outside to
avoid flickering

Add the translator comment

Readjust the bottom margins for two different headers

Wire in showPlanBenefits with trailMapExperiment

Only show the plan benefits header in the signup flow

Add some inline comments regarding potential improvements

Improve the selector specificity

Revert the comment that's added to the wrong place (facepalm)

southp commented:
Retested after rebasing and confirmed it works as described. I'm going ahead and deploying it, but follow-up comments are always welcome.
